lightning/src/ln/peers/handshake/mod.rs

```diff
@@ -0,0 +1,387 @@
+//! Handshake states can be advanced automatically, or by manually calling the appropriate step.
+//! Once complete, returns an instance of Conduit.
+
+use secp256k1;
+
+use bitcoin_hashes::{Hash, HashEngine};
+use bitcoin_hashes::sha256::Hash as Sha256;
+use secp256k1::{PublicKey, SecretKey};
+
+use ln::peers::{chacha, hkdf};
+use ln::peers::conduit::Conduit;
+use ln::peers::handshake::acts::{ActOne, ActThree, ActTwo};
+use ln::peers::handshake::hash::HandshakeHash;
+use ln::peers::handshake::states::{ActOneExpectation, ActThreeExpectation, ActTwoExpectation, HandshakeState};
+
+pub(crate) mod acts;
+mod hash;
+mod states;
+mod tests;
+
+/// Object for managing handshakes.
+/// Currently requires explicit ephemeral private key specification.
+pub struct PeerHandshake {
+	state: Option<HandshakeState>,
+	private_key: SecretKey,
+	ephemeral_private_key: SecretKey,
+
+	read_buffer: Vec<u8>,
+}
+
+impl PeerHandshake {
+	/// Instantiate a new handshake with a node identity secret key and an ephemeral private key
+	pub fn new(private_key: &SecretKey, ephemeral_private_key: &SecretKey) -> Self {
+		let handshake = PeerHandshake {
+			state: Some(HandshakeState::Blank),
+			private_key: (*private_key).clone(),
+			ephemeral_private_key: (*ephemeral_private_key).clone(),
+			read_buffer: Vec::new(),
+		};
+		handshake
+	}
+
+	/// Make the handshake object inbound in anticipation of a peer's first handshake act
+	pub fn make_inbound(&mut self) {
+		let public_key = Self::private_key_to_public_key(&self.private_key);
+		let (hash, chaining_key) = Self::initialize_state(&public_key);
+		self.state = Some(HandshakeState::AwaitingActOne(ActOneExpectation {
+			hash,
+			chaining_key,
+		}))
+	}
+
+	fn initialize_state(public_key: &PublicKey) -> (HandshakeHash, [u8; 32]) {
+		// do the proper initialization
+		let protocol_name = b"Noise_XK_secp256k1_ChaChaPoly_SHA256";
+		let prologue = b"lightning";
+
+		let mut sha = Sha256::engine();
+		sha.input(protocol_name);
+		let chaining_key = Sha256::from_engine(sha).into_inner();
+
+		let mut initial_hash_preimage = chaining_key.to_vec();
+		initial_hash_preimage.extend_from_slice(prologue.as_ref());
+
+		let mut hash = HandshakeHash::new(initial_hash_preimage.as_slice());
+		hash.update(&public_key.serialize());
+
+		(hash, chaining_key) // hash, chaining_key
+	}
+
+	/// Process act dynamically
+	/// The role must be set before this method can be called
+	pub fn process_act(&mut self, input: &[u8], remote_public_key: Option<&PublicKey>) -> Result<(Vec<u8>, Option<Conduit>, Option<PublicKey>), String> {
+		let mut response: Vec<u8> = Vec::new();
+		let mut connected_peer = None;
+		let mut remote_pubkey = None;
+
+		self.read_buffer.extend_from_slice(input);
+		let read_buffer_length = self.read_buffer.len();
+
+		match &self.state {
+			&Some(HandshakeState::Blank) => {
+				let remote_public_key = remote_public_key.ok_or("remote_public_key must be Some for outbound connections")?;
```

Comment:
in new_outbound in peer_handler.rs, the handshake is manually initiated because unlike new_inbound, that method immediately returns the first byte vector to send to the peer

Reply:
`remote_public_key` is passed as a parameter to `process_act` and it is only ever passed `None` in peer_handler.rs. Shouldn't you need to pass it `peer.their_node_id` instead?

Reply:
Normally, yes, but because new_outbound needs the response immediately, the initiate() method is called right there. It is the only act transition that needs the remote pubkey. I think once peer_handler is refactored, we will make new_inbound and new_outbound have more similar calling mechanisms, without one having a special condition for returning data.